op-batcher/batcher/channel_manager.go

```diff
@@ -139,9 +142,13 @@ func (s *channelManager) removePendingChannel(channel *channel) {
 // nextTxData dequeues frames from the channel and returns them encoded in a transaction.
 // It also updates the internal tx -> channels mapping
 func (s *channelManager) nextTxData(channel *channel) (txData, error) {
-	if channel == nil || !channel.HasTxData() {
+	if channel == nil {
+		s.log.Trace("no channel")
+		return txData{}, io.EOF
+	}
+	if !channel.HasTxData() {
 		s.log.Trace("no next tx data")
-		return txData{}, io.EOF // TODO: not enough data error instead
+		return txData{}, ErrInsufficientData
```

Comment:
There may be a few places where callers still assume that the error is `io.EOF`. I don't see error checks changed to the new error type, except for in tests. I'd prefer if we do this change to a custom error type in a separate PR, or is it necessary for the scope of this PR?

Reply:
The motivation was to be able to differentiate the two different cases which return `io.EOF` to make testing easier. But it is not strictly necessary for this PR [510464d](https://github.com/ethereum-optimism/optimism/pull/12002/commits/510464d28cda1e5f0c2690be810ddda5c24a68c0)

Reply:
The driver does behave differently (logging behavior) with the change, as the error bubbles up: 

https://github.com/ethereum-optimism/optimism/blob/510464d28cda1e5f0c2690be810ddda5c24a68c0/op-batcher/batcher/driver.go#L536-L547

Reply:
The driver deliberately checks `io.EOF` here and only trace- instead of error-logs, because it's an expected sentinel error in the normal path of execution. So error logging on `ErrInsufficientData` would be wrong. Also wondering why, from the perspective of the caller of `TxData`, would it ever behave differently depending on `io.EOF` or `ErrInsufficientData`? It shouldn't care whether there's no data because there's no channel or because the current channel doesn't yield a frame yet.

Reply:
Agreed!